feat: Blob filestores

mrzeszutko opened the pull request:
Added `FileStoreBlobClient` that enables reading and writing blobs to cloud storage (S3, GCS) or local filesystem. This provides an alternative blob source alongside the existing blob sink server, L1 consensus layer, and archive (e.g., Blobscan).

## Blob Download Logic

The download strategy varies based on sync state:

- **Historical sync:** Blob sink → FileStore → L1 consensus → Archive
- **Near-tip sync:** Blob sink → FileStore → L1 consensus → FileStore (with retries) → Archive

Near-tip sync includes retries with exponential backoff, assuming filestores should already contain the data when blobs are queried.

## Blob Upload Logic

Blobs are uploaded only if the node has write access to a filestore. Uploads occur in three scenarios:

1. `SequencerPublisher` publishes blobs when posting a checkpoint to L1
2. The archiver fetches blobs from any source
3. A validator receives a valid block proposal

## Configuration

The following environment variables are introduced:

| Variable | Description |
|----------|-------------|
| `BLOB_FILE_STORE_URLS` | Comma-separated URLs for reading blobs (`s3://`, `gs://`, `file://`, `https://`) |
| `BLOB_FILE_STORE_UPLOAD_URL` | URL for uploading blobs (`s3://`, `gs://`, `file://`) |

`BLOB_FILE_STORE_URLS` can also be defined in the network configuration, enabling centralized management of these URLs across nodes.

mrzeszutko committed:
Blob filestores

PhilWindle commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -532,3 +672,12 @@ function getBeaconNodeFetchOptions(url: string, config: BlobSinkConfig, l1Consen
 }
+
+/** Fisher-Yates shuffle for load distribution across filestores */
+function shuffleArray<T>(array: T[]): T[] {
```
This would be great to add in foundation.

spalladino replied:
We already have it, see `foundation/src/array/array.ts`

mrzeszutko replied:
I'll switch to the existing implementation

PhilWindle commented on yarn-project/validator-client/src/validator.ts:
```diff
@@ -345,6 +351,20 @@ export class ValidatorClient extends (EventEmitter as new () => WatcherEmitter)
     this.metrics.incSuccessfulAttestations(inCommittee.length);
 
+    // Upload blobs to filestore after successful re-execution (fire-and-forget)
+    if (validationResult.reexecutionResult?.block && this.fileStoreBlobUploadClient) {
```
I think this part is probably going to need to change isn't it @spalladino? This only works today as it's 1 block per checkpoint.

spalladino replied:
I can take care of it once I move on to the validator client.

spalladino commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -1,33 +1,50 @@
+    this.fileStoreUploadClient = opts.fileStoreUploadClient;
+
+    if (this.fileStoreUploadClient && !opts.onBlobsFetched) {
+      this.opts.onBlobsFetched = blobs => {
+        this.uploadBlobsToFileStore(blobs);
+      };
```
Nit: avoid mutating arguments if possible. This changes the `opts` object created by the caller.

mrzeszutko replied:
It does but it is executed only if the caller did not provide the handler: `if (this.fileStoreUploadClient && !opts.onBlobsFetched) {`

spalladino commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -296,6 +401,41 @@ export class HttpBlobSinkClient implements BlobSinkClientInterface {
+   * @param fillResults - Callback to fill in results
+   * @param ctx - Logging context
+   */
+  private async tryFileStores(
+    blobHashes: Buffer[],
+    fillResults: (blobs: BlobJson[]) => BlobWithIndex[],
+    ctx: { blockHash: string; blobHashes: string[]; indices?: number[] },
+  ): Promise<void> {
+    // Shuffle clients for load distribution
+    const shuffledClients = shuffleArray([...this.fileStoreClients]);
+
+    for (const client of shuffledClients) {
+      try {
+        const blobHashStrings = blobHashes.map(h => `0x${h.toString('hex')}`);
+        this.log.trace(`Attempting to get ${blobHashStrings.length} blobs from filestore`, {
+          url: client.getBaseUrl(),
+          ...ctx,
+        });
+        const blobs = await client.getBlobsByHashes(blobHashStrings);
+        if (blobs.length > 0) {
+          const result = fillResults(blobs);
+          this.log.debug(`Got ${blobs.length} blobs from filestore (total: ${result.length}/${blobHashes.length})`, {
+            url: client.getBaseUrl(),
+            ...ctx,
+          });
+        }
+      } catch (err) {
+        this.log.warn(`Failed to fetch from filestore: ${err}`, { url: client.getBaseUrl() });
+      }
+    }
```
I may be missing something, but aren't we missing an early return here once we've collected all blobs? Also, if we are looking for two blobs and we got one from the first filestore, I think we keep requesting the original two from the next ones, since `blobHashes` does not get updated.

mrzeszutko replied:
you're right - I missed this and will fix

spalladino commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -237,12 +321,33 @@ export class HttpBlobSinkClient implements BlobSinkClientInterface {
+          `Waiting ${delayMs}ms before filestore retry attempt ${attempt + 1}/${fileStoreMaxRetries}`,
+          ctx,
+        );
+        await sleep(delayMs);
```
There's a `retry` function in `foundation` that may come in handy here.

mrzeszutko replied:
For simplicity I'll switch to using `retry` function for now - I'll remove additional config values `BLOB_FILE_STORE_MAX_RETRIES` and `BLOB_FILE_STORE_RETRY_BASE_MS` as the intervals in `retry` are set in seconds

spalladino commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -188,6 +253,14 @@ export class HttpBlobSinkClient implements BlobSinkClientInterface {
+    // Fire callback when returning blobs (fire-and-forget)
+    const returnWithCallback = (blobs: BlobWithIndex[]): BlobWithIndex[] => {
+      if (blobs.length > 0 && this.opts.onBlobsFetched) {
+        void Promise.resolve().then(() => this.opts.onBlobsFetched!(blobs));
```
Stupid question: why not just `void this.opts.onBlobsFetched(blobs)`?

mrzeszutko replied:
This changes the order of execution (void this.opt.onBlobsFetched will be executed right away and void Promise.resolve().then() fully async - will get queued on the microtask queue). To be honest in this case it probably does not matter at all and both will work correctly but my intention was to first return the data to the caller, then execute the handler async

spalladino commented on yarn-project/blob-sink/src/filestore/filestore_blob_client.ts:
```diff
@@ -0,0 +1,133 @@
+  async saveBlobs(blobs: Blob[], skipIfExists = true): Promise<void> {
+    for (const blob of blobs) {
+      await this.saveBlob(blob, skipIfExists);
+    }
```
Can we run these in parallel? Either a `Promise.all`, or an `asyncPool` if we think we may try pushing too many at once.

mrzeszutko replied:
I'll use Promise.all in this case + merge with the other method mentioned below

spalladino commented on yarn-project/blob-sink/src/filestore/filestore_blob_client.ts:
```diff
@@ -0,0 +1,133 @@
+  async saveBlobsFromL1(blobs: BlobWithIndex[], skipIfExists = true): Promise<void> {
+    for (const blobWithIndex of blobs) {
+      await this.saveBlob(blobWithIndex.blob, skipIfExists);
+    }
```
Same as above. Also, let's try to merge with the fn above to avoid duplication?

mrzeszutko replied:
I'll use Promise.all in this case + merge with the other method mentioned above

spalladino commented on yarn-project/blob-sink/src/filestore/filestore_blob_client.ts:
```diff
@@ -0,0 +1,133 @@
+   */
+  testConnection(): Promise<boolean> {
+    // This implementation will be improved in a separate PR
+    // Currently underlying filestore implementations do not expose an easy way to test connectivitiy
```
Curious to hear your thoughts on how to test for each filestore

mrzeszutko replied:
I still need to think it through but for S3 we could try [HeadBucketCommand](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3/command/HeadBucketCommand/), for GCS probably bucket.exists or bucket.getMetadata, local filesystem seems easy as we can just check if the folder exists

alexghr commented on yarn-project/blob-sink/src/filestore/filestore_blob_client.ts:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    // index=0 is deliberate as we don't know the actual blob index in most cases when the filestores are used (blobs are saved in the filestores before we ever fetch them from L1)
+    const json = blob.toJson(0);
```
Should this use a different index than 0 to represent "unknown" in this case?

mrzeszutko replied:
I'll change this to -1

alexghr commented on yarn-project/blob-sink/src/filestore/filestore_blob_client.ts:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    // index=0 is deliberate as we don't know the actual blob index in most cases when the filestores are used (blobs are saved in the filestores before we ever fetch them from L1)
+    const json = blob.toJson(0);
```
Should this use a different index than 0 to represent "unknown" in this case?

mrzeszutko replied:
I'll change this to -1

alexghr commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -40,6 +57,20 @@ export class HttpBlobSinkClient implements BlobSinkClientInterface {
+
+    this.fileStoreUploadClient.saveBlobsFromL1(blobs, true).catch(err => {
+      this.log.warn(`Failed to upload ${blobs.length} blobs to filestore`, err);
+    });
```
I think this needs a `void` in the beginning to ignore the promise otherwise the linter/ts will complain

mrzeszutko replied:
Good catch - I'll add void

alexghr commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -296,6 +401,41 @@ export class HttpBlobSinkClient implements BlobSinkClientInterface {
+
+    for (const client of shuffledClients) {
+      try {
+        const blobHashStrings = blobHashes.map(h => `0x${h.toString('hex')}`);
+        this.log.trace(`Attempting to get ${blobHashStrings.length} blobs from filestore`, {
+          url: client.getBaseUrl(),
+          ...ctx,
+        });
+        const blobs = await client.getBlobsByHashes(blobHashStrings);
```
I have a question about this + `fillResults` (since it's not part of the diff):

If I need to pull three blobs with hashes A, B, C and I get blob B from the first file store, will the second file store only be asked about blobs A and C?

mrzeszutko replied:
This will get fixed - @spalladino already found that this was implemented incorrectly, it will work as you described

alexghr commented on yarn-project/blob-sink/src/client/http.ts:
```diff
@@ -113,34 +160,49 @@ export class HttpBlobSinkClient implements BlobSinkClientInterface {
+        const res = await this.fetch(`${this.config.blobSinkUrl}/blobs`, {
+          method: 'POST',
+          headers: { 'Content-Type': 'application/json' },
+          body: JSON.stringify({
+            // Snappy compress the blob buffer
+            blobs: blobs.map((b, i) => ({ blob: outboundTransform(b.toBuffer()), index: i })),
+          }),
```
THis had me super confused because it does snappy compression. I thought it was new code and the PR missed decompressing when retrieving blobs, but this is part of the old blob sink implementation.

mrzeszutko replied:
Old code does not need to decompress as `outboundTransform` is used when sending to BlobSink and `inboundTransform` is used by BlobSinkServer when saving data. When the data is downloaded there is no compression at all. Great catch - I think it might be worth to use compression to limit the amount of data that's stored and transferred

mrzeszutko replied:
I added snappy compression to blob upload/download to/from filestores

mrzeszutko commented:
@spalladino @alexghr @PhilWindle PR ready for re-review. Most notable changes:
- I removed `BLOB_FILE_STORE_MAX_RETRIES` and `BLOB_FILE_STORE_RETRY_BASE_MS` settings and instead am using existing `retry` function with hardcoded backoff intervals of [1, 1, 2] seconds
- I changed the blob index to -1 for filestores
- I added snappy compression 
- I fixed the issue with querying filestores only with missing hashes
